### PR TITLE
[WIP] Serialize local functions inside transparent functions

### PR DESCRIPTION
**Pull request by slavapestov**

### CODE REVIEW

**Comment by slavapestov** [comment]
@swift-ci Please test

**Comment by slavapestov** [comment]
@swift-ci Please test

File: stdlib/public/runtime/Reflection.mm [hunk 1]
```diff
@@ -497,7 +486,18 @@ static Mirror reflect(HeapObject *owner,
...
   
+SWIFT_RUNTIME_STDLIB_INTERFACE
+extern "C"
+intptr_t swift_TupleMirror_count(HeapObject *owner,
+                                 const OpaqueValue *value,
+                                 const Metadata *type) {
+  auto Tuple = static_cast<const TupleTypeMetadata *>(type);
+  return Tuple->NumElements;
+}
+
```
Comment (jckarter) [inline stdlib/public/runtime/Reflection.mm]: Was this reordering intentional?

File: lib/SIL/SILDeclRef.cpp [hunk 2]
```diff
@@ -308,25 +280,26 @@ bool SILDeclRef::isClangGenerated() const {
...
+      return SILLinkage::Shared;
     moduleContext = moduleContext->getParent();
   }
   
   // Currying and calling convention thunks have shared linkage.
   if (isThunk())
-    // If a function declares a @_cdecl name, its native-to-foreign thunk is
-    // exported with the visibility of the function.
+    // If a function declares a @_silgen_decl name, its native-to-foreign thunk
+    // is exported with the visibility of the function.
```
Comment (jckarter) [inline lib/SIL/SILDeclRef.cpp]: The comment was correct before.

File: stdlib/public/runtime/Reflection.mm (hunk 1 above)
Comment (slavapestov) [inline stdlib/public/runtime/Reflection.mm]: Yeah, it is in its own patch. Unless I'm missing something the old ordering was a mistake?

File: lib/SIL/SILDeclRef.cpp (hunk 2 above)
Comment (slavapestov) [inline lib/SIL/SILDeclRef.cpp]: ok

File: stdlib/public/runtime/Reflection.mm (hunk 1 above)
Comment (jckarter) [inline stdlib/public/runtime/Reflection.mm]: Why does it matter?

Comment (slavapestov) [inline stdlib/public/runtime/Reflection.mm]: It feels more natural to put the tuple functions together, after the '// -- Tuple destructuring.' section. Previously swift_TupleMirror_count came first, then some other unrelated functions, then the rest of the TupleMirror stuff.

If you prefer I can drop this patch, it's not related to anything else in the series. Arguably moving code around destroys 'git blame' history, but I have some other changes to this file that are not yet committed, so I decided to fix this nit.

Comment (jckarter) [inline stdlib/public/runtime/Reflection.mm]: Nah, it's fine. Just wondering why.

Comment (slavapestov) [inline stdlib/public/runtime/Reflection.mm]: It's so much harder to make your code look good when the language doesn't enforce bottom-up function order ;)

**Comment by slavapestov** [comment]
@swift-ci Please test

File: stdlib/public/runtime/Reflection.mm (hunk 1 above)
Comment (jckarter) [inline stdlib/public/runtime/Reflection.mm]: Yeah, the generative approach to declarations in ML family languages is so much more elegant than this multi-pass declarative crap

**Comment by slavapestov** [comment]
@swift-ci Please test

**Comment by slavapestov** [comment]
@swift-ci Please test

**Comment by slavapestov** [comment]
@swift-ci Please test

**Comment by slavapestov** [comment]
@swift-ci Please test

**Comment by slavapestov** [comment]
@swift-ci Please test

**Comment by slavapestov** [comment]
@swift-ci Please test

**Comment by slavapestov** [comment]
Note to self:

```
    Swift :: 1_stdlib/BridgeNonVerbatim.swift
    Swift :: 1_stdlib/Builtins.swift
    Swift :: 1_stdlib/RuntimeObjC.swift
    Swift :: 1_stdlib/SpriteKit.swift
    Swift :: DebugInfo/specialization.swift
    Swift :: IRGen/enum_derived.swift
    Swift :: IRGen/runtime_calling_conventions.swift
    Swift :: IRGen/vector_reduction.swift
    Swift :: SIL/Serialization/shared_function_serialization.sil
    Swift :: SIL/Serialization/specializer_can_deserialize.swift
    Swift :: SIL/parse_stdlib_0.sil
    Swift :: SIL/parse_stdlib_1.sil
    Swift :: SIL/parse_stdlib_10.sil
    Swift :: SIL/parse_stdlib_11.sil
    Swift :: SIL/parse_stdlib_12.sil
    Swift :: SIL/parse_stdlib_13.sil
    Swift :: SIL/parse_stdlib_14.sil
    Swift :: SIL/parse_stdlib_15.sil
    Swift :: SIL/parse_stdlib_16.sil
    Swift :: SIL/parse_stdlib_2.sil
    Swift :: SIL/parse_stdlib_3.sil
    Swift :: SIL/parse_stdlib_4.sil
    Swift :: SIL/parse_stdlib_5.sil
    Swift :: SIL/parse_stdlib_6.sil
    Swift :: SIL/parse_stdlib_7.sil
    Swift :: SIL/parse_stdlib_8.sil
    Swift :: SIL/parse_stdlib_9.sil
    Swift :: SILOptimizer/DestructorAnalysis.swift
    Swift :: SILOptimizer/allocbox_to_stack.sil
    Swift :: SILOptimizer/bridged_casts_folding.swift
    Swift :: SILOptimizer/cast_folding_objc.swift
    Swift :: SILOptimizer/cast_folding_objc_no_foundation.swift
    Swift :: SILOptimizer/devirt_base_class.swift
    Swift :: SILOptimizer/devirt_default_case.swift
    Swift :: SILOptimizer/diagnostic_constant_propagation.swift
    Swift :: SILOptimizer/eager_specialize.sil
    Swift :: SILOptimizer/functionsigopts.sil
    Swift :: SILOptimizer/functionsigopts_self.swift
    Swift :: SILOptimizer/functionsigopts_sroa.sil
    Swift :: SILOptimizer/prespecialize.swift
    Swift :: SILOptimizer/spec_conf1.swift
    Swift :: SILOptimizer/spec_conf2.swift
    Swift :: SILOptimizer/spec_recursion.swift
    Swift :: SILOptimizer/specialize.sil
    Swift :: SILOptimizer/specialize_chain.swift
    Swift :: SILOptimizer/specialize_default_witness.sil
    Swift :: SILOptimizer/specialize_ext.swift
    Swift :: SILOptimizer/specialize_partial_apply.swift
    Swift :: SILOptimizer/specialize_recursive_generics.sil
    Swift :: SILOptimizer/specialize_unconditional_checked_cast.swift
    Swift :: Serialization/basic_sil.swift
    Swift :: stdlib/ArrayNew.swift.gyb
    Swift :: stdlib/Dictionary.swift
    Swift :: stdlib/Set.swift
```

**Pull request closed without merging**